README.md
# 41934-Advanced-BIM-Group-22

Please follow instructions provided in the main.py file for it to run as intended. It should be able to be run in both Blender with blenderbim and just in the command prompt (CMD) if at least python 3.7 is installed


A2_FutureBIM/CHANGES.md
CHANGES MADE:

For this assignment about FutureBIM we focused on linking the first assignment to this one in terms of being able to caluclate the ventilation rates required to uphold a proper indoor climate accordingly to thresholds from DS/EN16798. 
In the first assignment OpenBIM we used the IFC model alongside with BLENDER to get an insight of how we could approach the problem and it helped us to develop code that could extract the data needed. 
This is how we linked these two assignments together and what we focused on here was to further improve the code for future use and make it more visually available/pleasing by "hacking" the html code.
To run the results simply open up the index.html from the file folder path: HTML-Build-IFC-Converter-main\output\duplex\index.html (preferably with google chrome)

To achieve our goal we needed to make changes to the following four files:
 * HTMLBuild.py (inside of folder: HTML-Build-IFC-Converter-main)
 * html-build.css (inside of folder: HTML-Build-IFC-Converter-main\output\css)
 * index.html (inside of folder: HTML-Build-IFC-Converter-main\output\duplex)
 * html-build.js (inside of folder: HTML-Build-IFC-Converter-main\output\js)
  
The changes vary a lot with everything from small comestic changes to data extraction and to make it easier to identify where code has been changed/made documentation is provided and explained.
The changes in no particular order:

 1. Changed the color from being yellow to blue (line 4-27 in html-build.css)
 2. Added a HTML titel to the index (line 240-246 in html-build.css)
 3. Added a titel to the properties box in the bottom right (line 162-175 in html-build.css)
 4. Changed the general font. (Throughout html-build.css)
 5. In the HTMLBuild.py we added a loop for extracting the information in the bottom left box about spaces. (line 135-159 in HTMLBuild.py)
 6. Added table to display the extracted information from point 5. on the left half of the screen below the building (line 248-258 in html-build.css)
 7. Added longitude and latitude to the properties box (line 16-22 in html-build.js)
 8. Rounded off the elevation to one decimal (line 11-16 in index.html)
 9. Fixed sizes of boxes and allignment to be more symmetrical (line 32-45 in html-build.css)
 10. Changed text from "happy" to "Click on a floor to see the attributes" (line 26 in html-build.js)
 
Future work: 
If we had more time we wanted to implement code that would then take areas of the spaces and then calculate how much ventilation rate is needed to reach each of the the indoor environment classes from 1 to 4 from DS/EN16798
and display it on the html page. 
 

A2_FutureBIM/HTMLBuild.py
''' written by Tim McGinley 2022 '''

import ifcopenshell
import os.path
import time

def modelLoader(name):

    ''' 
        load the IFC file 
    '''
    
    model_url = "model/"+name+".ifc"
    start_time = time.time()

    if (os.path.exists(model_url)):
        model = ifcopenshell.open(model_url)
        print("\n\tFile    : {}.ifc".format(name))
        print("\tLoad    : {:.2f}s".format(float(time.time() - start_time)))
        
        start_time = time.time()
        writeHTML(model,name)
        print("\tConvert : {:.4f}s".format(float(time.time() - start_time)))
        
    else:
        print("\nERROR: please check your model folder : " +model_url+" does not exist")

def writeHTML(model,name):

    ''' 
        write the HTML entities 
    '''
    
    # parent directory - put in setting file?
    parent_dir = "output/"
    # create an HTML file to write to
    if (os.path.exists("output/"+name))==False:
        path = os.path.join(parent_dir, name)
        os.mkdir(path)
    
    f_loc="output/"+name+"/index.html"
    f = open(f_loc, "w")
    cont=""
    
    # ---- START OF STANDARD HTML
    cont+=0*"\t"+"<html>\n"
    # ---- ADD HEAD
    cont+=1*"\t"+"<head>\n"
    # ---- ADD HTMLBUILD CSS - COULD ADD OTHERS HERE :)
    cont+=2*"\t"+"<link rel='stylesheet' href='../css/html-build.css'></link>\n"
    # ---- ADD HTMLBUILD CSS - COULD ADD OTHERS HERE :)
    cont+=2*"\t"+"<link rel='stylesheet' href='https://cdn.jsdelivr.net/npm/bootstrap@3.4.1/dist/css/bootstrap.min.css' integrity='sha384-HSMxcRTRxnN+Bdg0JdbxYKrThecOKuH5zCYotlSAcp1+c8xmyTe9GYg1l9a69psu' crossorigin='anonymous'>"
    # ---- ADD HTMLBUILD JS - COULD ADD OTHERS HERE :)
    cont+=2*"\t"+"<script src='../js/html-build.js'></script>\n"
    # ---- JQUERY - IT WOULD BE CRAZY NOT TO
    cont+=2*"\t"+"<script src='https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.js'></script>\n"
    # ---- CLOSE HEAD
    cont+=1*"\t"+"</head>\n"
    # ---- ADD BODY
    cont+=1*"\t"+"<body onload=\"main()\">\n"  
    # ---- ADD TITEL
    cont+=1*"\t"+"<h1 id='projectTitle'></h1>"
    
    # ---- ADD CUSTOM HTML FOR THE BUILDING HERE
    cont+=writeCustomHTML(model)
    
    # ---- CLOSE BODY AND HTML ENTITIES
    cont+=1*"\t"+"</body>\n"   
    cont+=0*"\t"+"</html>\n"

    # ---- WRITE IT OUT
    f.write(cont)
    f.close()

    # ---- TELL EVERYONE ABOUT IT
    print("\tSave    : "+f_loc)

def writeCustomHTML(model):

    ''' 
        write the custom HTML entities 
    '''
    
    custom=""
    site_elev = 0 # variable for to store the elevation of the site
    
    # ---- DEFINE THE MODEL
    
    custom+=2*"\t"+"<model->\n"
    
    # ---- ADD PROJECT CUSTOM ENTITY
    project = model.by_type('IfcProject')[0]
    custom+=3*"\t"+"<project- name=\"{d}\">\n".format(d=project.LongName) # ---- Removed 8?
    
    # it looks like it would make sense to use the DOM here and append stuff to it...
    
   
        
        
    
    
    
    
    # ---- ADD SITE CUSTOM ENTITY
    site = model.by_type('IfcSite')[0]
    site_elev = site.RefElevation
    custom+=4*"\t"+"<site- lat=\"{}\" long=\"{}\" elev=\"{}\">\n".format(site.RefLatitude,site.RefLongitude,site_elev )

    # ---- ADD BUILDING CUSTOM ENTITY
    custom+=5*"\t"+"<building->\n"
    
   
    
    # ---- ADD FLOOR CUSTOM ENTITIES
    floors = model.by_type('IfcBuildingStorey')
    floors.sort(key=lambda x: x.Elevation, reverse=True)
   
    
    # ---- CLASSIFY THE FLOORS AS LOWER, GROUND OR UPPER AND WRITE TO CUSTOM ENTITIES
    custom+= classifyFloors(floors,site_elev)
    
   
    
    
    # ---- CLOSE BUILDING
    custom+=5*"\t"+"</building->\n"
    
    
    
    # ---- CLOSE SITE AND PROJECT
    custom+=4*"\t"+"</site->\n"
    custom+=3*"\t"+"</project->\n"
      # ---- ADD SPACE CUSTOM ENTITY
    
#---------------------------------------------------------------------------------------------------------------------------------------------------#    
    # HERE GROUP 22 MADE OWN CODE TO GET THE LEVEL, AREA AND SPACEE LONGNAME OF SPACES IN THE BUILDING
    custom+=4*"\t"+"<spaces-> <h4>Spaces</h4> <table><th>Space</th><th>Level</th><th>Area</th>\n"
    spaces = model.by_type('IfcSpace')
    for space in spaces: # ---- FOR EVERY SPACE, GIVE THEM NEW NAMES
        space_LongName = space.LongName
        space_name = space.Name
        for definition in space.IsDefinedBy: # ---- FIND THE PROPERTIES FOR EACH SPACES
            if definition.is_a('IfcRelDefinesByProperties'):
                property_set = definition.RelatingPropertyDefinition
                if property_set.Name == "PSet_Revit_Dimensions": # ---- IF THE PROPERTY IS DIMENSIONS
                    for property in property_set.HasProperties:
                        if property.Name == "Area":  # ---- SAVE THE AREA
                            space_area = property.NominalValue.wrappedValue
                if property_set.Name == "PSet_Revit_Constraints": # ---- IF THE PROPERTY IS CONSTRAINTS
                    for property in property_set.HasProperties:
                        if property.Name == "Level": # ---- SAVE LEVEL OF THE SPACE
                            space_level = property.NominalValue.wrappedValue
        
        # ---- SAVE AS AN ARRAY AND SAVES INTO THE "CUSTOM STRING" TO BE ABLE TO DISPLAY THE DATA INSIDE THE TABLE OF THE FRONT PAGE
        #custom+=5*"\t"+"<space- longName=\'{}\' name=\'{}\' area=\'{}\' level=\'{}\'>{}, {}, Area: {}m<sup>2</sup> <br>\n".format(space_LongName,space_name,space_area,space_level,space_LongName, space_level, round(float(space_area),1))
        custom+=5*"\t"+"<tr><td>{}</td><td>{}</td><td>{}m<sup>2</sup></td></tr>\n".format(space_LongName,space_level,round(float(space_area),1))
        #custom+=5*"\t"+"</space->\n"
    custom+=4*"\t"+"</table></spaces->\n"
#---------------------------------------------------------------------------------------------------------------------------------------------------#  


    # ---- END OF MODEL ENTITY
    custom+=2*"\t"+"</model->\n"
    
    # ---- ADD VIEWS.
    custom+=2*"\t"+"<view->\n"
    
    # ---- ADD PLAN.
    custom+=3*"\t"+"<plan-></plan->\n"
    
    # ---- ADD PROPERTIES ETC.
    custom+=3*"\t"+"<props-></props->\n"
    
    # ---- CLOSE VIEWS
    custom+=2*"\t"+"</view->\n"
    
    # ---- RETURN THE CUSTOM HTML
    return custom

def classifyFloors(floors,site_elev):

    '''
    another way after arranging them would be to split them into above and below ground floor sets.
    '''
    
    floor_entities = ''
    
    # these are interesting and probably should be output somwhere - maybe to the building data?
    lower_floors = sum(f.Elevation < 0.1 for f in floors)
    level = len(floors)-lower_floors
    
    for floor in floors:
        # check if floor is lower than elevation...
        type = "floor_upper"
        if ( site_elev-.1 <= floor.Elevation <= site_elev+.1):
            type = "floor_ground"
        elif (site_elev < floor.Elevation):
            type = "floor_upper"
        else:
            type = "floor_lower"
           
        # THE SPAN STUFF SHOULD BE DEALT WITH IN JS...
        
        floor_entities+=6*"\t"+"<floor- class=\""+type+"\" name='{}'  level='{}' elev=\"{}\" >{}<span class=\"floor_stats\">{}</span> </floor->\n".format(floor.Name, level, floor.Elevation,floor.Name, round(float(floor.Elevation),3))     
        level-=1
        if (type == "floor_ground"):
            floor_entities+=6*"\t"+"<ground-></ground->\n"
            
    return floor_entities


A2_FutureBIM/output/js/html-build.js
/* written by Tim McGinley 2022 */

// ok in here we need to include a lot of stuff.
// we need a menu... where would this fit?
// we need to start (over)loading stuff into the DOM.
// we need to split the screen into section and plan and KPIs and info - where should this go?


function main() {
	
	// calculate the floors
	const floors = document.getElementsByTagName("floor-");
	let num_floors = floors.length;
	console.log(num_floors);
	
	// add data to the properties box // added more boxes for the extra information we want to display
	$('props-').prepend('Number of floors: '+num_floors);
	$('props-').prepend('Site elevation: '+$('site-').attr('elev')+'<br>');
	$('props-').prepend('Site latitude: '+$('site-').attr('lat')+'<br>');
	$('props-').prepend('<h4>Properties</h4>Site longitude: '+$('site-').attr('long')+'<br>');
	
	$('h1').prepend($('project-').attr('name')+'<br>');
	
	
	// load the plan so we can edit it and changed text from happy
	plan('Click on a floor to see the attributes');
	
	// The .each() method is unnecessary here:
	$( 'floor-' ).each(function() {
	console.log($(this)[0].innerHTML);
		$( this ).on("click", function(){
			//$('plan-').css("background-color","black");
															// add more data we extract
			changePlan('Floor name: ' + $(this).attr('name') + '<br>Level: ' + $(this).attr('level') + '<br>Elevation: ' + $(this).attr('elev'));
			//$( this ).innerHTML
		});
	});
	
	
}

function plan(text) {
jQuery('<div>', {
    id: 'plan',
    class: 'plan',
    title: 'click a floor to see the plan',
	html:text
}).appendTo('plan-');  
	
}

function changePlan(text) {
	$('#plan').html(text);
}





// <project-> - title etc.... | <site-> - also menu? site specific data?
// ---------------------------------------------------------------------
// <building-> - name of the building? this then needs to split in two...
// could also be more views and show a 3d? but maybe left is consistent
// ---------------------------------------------------------------------
// #section                   |               #plan
// this shows the floors      |      this shows a floor in plan         |
// in section                 |                 #plan                   |
//    <floor...->              -----------------------------------------
//                            |                 <properties->           |
// ---------------------------------------------------------------------


A4_FinalProject/main.py
#ONLY TESTED WITH PYTHON 3.7
#ONLY TESTED WITH IFCOPENSHELL FOR PYTHON 3.7
#Please operate in Comand Prompt

#https://pypi.org/project/install-ifcopenshell-python/0.1.0/
#Really important to have ifcopenshell or else it is not possible to work with the IFC formats.
import ifcopenshell
#.geom includes a library that helps when working with geometry
import ifcopenshell.geom
from ifcopenshell import geom
#pip install XlsxWriter. XlsxWriter is the package responsible to create and handle excel
import xlsxwriter
#os is used to handle local paths
import os
#pip install numpy. Numpy is needed in order to work with arrays, matrixes and vectors.
import numpy as np



#only run this if trying to run the code inside of blender
#------------------------------------------------------------#
#import bpy
#from blenderbim.bim.ifc import IfcStore
#model = IfcStore.get_file()
#------------------------------------------------------------#


#Defines the model to the desired ifc file. Please put the exact model in the folder called "model"
model = ifcopenshell.open('model/Duplex_A_20110907.ifc')


#Please make a folder called "output" in the folder or else no excel sheet can be created 
workbook = xlsxwriter.Workbook('output/Ventilation and solar radiation.xlsx')


#Makes it able to make text bold inside of excel
bold = workbook.add_format({'bold': True})


#This is used to get find the types inside of the ifc file. Its getting used later on
spaces = model.by_type("IfcSpace")
windows = model.by_type("IfcWindow")



#makes spreadsheet in excel. This uses the xlsxwriter package
def makeASheet(ifcType):
        sheet = workbook.add_worksheet('Spaces')
        #writes into a sheet. (row, coloum, data)
        sheet.write(0,0,'Name:',bold)
        sheet.write(0,1,'Number:',bold)
        sheet.write(0,2,'Level:',bold)
        sheet.write(0,3,'Area [m2]:',bold)
        sheet.write(0,4,'IEQ Class:',bold)
        sheet.write(0,5,'Ventilation rate [l/s]:',bold)
        
        #fresh start for the upcomming code, working as a reset
        row =1
        column =1
        for entity in model.by_type(ifcType):
            #this writes the data to the sheet
            #sheet.write(LongName is the actual name and Name is the number linked to the name)
            #float is used so excel knows is a number and not text
            sheet.write(row,0,str(entity.LongName))
            sheet.write(row,column,str(entity.Name))  
            sheet.write(row,column+3,float(2)) 
            #this 'iterates' row so that each time we step down a row.
            #otherwise each new entry would overwrite the previous entry.
            row+=1  

        #reseting the row and column for the upcomming code
        row =1
        column =1
        
        #Gathering information from the IFC file, in this case the what is within spaces
        for space in spaces:
            for definition in space.IsDefinedBy:
                
                ## To support IFC2X3, we need to filter our results.
                if definition.is_a('IfcRelDefinesByProperties'):
                    property_set = definition.RelatingPropertyDefinition
                    
                    
                    #Sort by the name of the propertySet and search for the area within the spaces. This loops until all spaces are found.
                    if property_set.Name == "PSet_Revit_Dimensions":
                       for property in property_set.HasProperties:
                            if property.Name == "Area":
                                #finds the value and rounds it to one decimal
                                area = property.NominalValue.wrappedValue
                                area = round(area,1)
                                sheet.write(row,column+2,area)
                                #1.4 is the ventilation rate according to DS 16798 table B.6
                                sheet.write(row,column+4,area*1.4)
                                row+=1

                    #sort by the name of the property
                    if property_set.Name == "PSet_Revit_Constraints":
                        for property in property_set.HasProperties:
                            if property.Name == "Level":
                                space_level = property.NominalValue.wrappedValue
                                sheet.write(row,column+1,space_level)
                                
#end of the loop. Makes a sheet called IfcSpace in which all the information will be stored and showcased in the excel sheet.
makeASheet('IfcSpace')


#makes an empty list and then fills it with the IfcWindow data inside of the ifcmodel
productList = []
for product in model.by_type("IfcWindow"):
    productList.append(product.is_a())
productList = list(set(productList))
productList.sort()

#Functions To define the matrix placement of the window objects.
def a2p(o,z,x):
    y = np.cross(z, x) 
    r = np.eye(4) 
    r[:-1,:-1] = x,y,z 
    r[-1,:-1] = o 
    return r.T
    
def axis2placement(plc): 
    z = np.array(plc.Axis.DirectionRatios if plc.Axis else (0,0,1)) 
    x = np.array(plc.RefDirection.DirectionRatios if plc.RefDirection else (1,0,0)) 
    o = plc.Location.Coordinates 
    return a2p(o,z,x) 
    
def local_placement(plc):
    if plc.PlacementRelTo is None: 
        parent = np.eye(4)
    else:
        parent = local_placement(plc.PlacementRelTo)
    return np.dot(parent, axis2placement(plc.RelativePlacement))


#now makes a new sheet as previously.
def makeASheet(ifcType):
        sheet = workbook.add_worksheet("Windows")
        sheet.write(0,0,'Name:',bold)
        sheet.write(0,1,'Level:',bold)
        sheet.write(0,2,'Orientation:',bold)
        sheet.write(0,3,'Height[m]:',bold)
        sheet.write(0,4,'Width[m]:',bold)
        sheet.write(0,5,'Area [m2]:',bold)
        sheet.write(0,6,'Solar radiation [W]:',bold)
        arealist = []
        row =1
        column =1
        
        #loops the code and writes all the entity names one row below each other
        for entity in model.by_type(ifcType):
            sheet.write(row,0,str(entity.Name))
            row+=1
       
        #reseting row and coloumns for the upcomming data   
        row =1
        column =1
        
        
        #Gathering information from the IFC file, in this case the windows
        for window in windows:
            #get property sets attached to windows through IsDefinedBy"
            for definition in window.IsDefinedBy:
                
                #To support IFC2X3, we need to filter our results.
                if definition.is_a('IfcRelDefinesByProperties'):
                    property_set = definition.RelatingPropertyDefinition
                    
                    
                    #Sort by the name of the propertySet and gets the height of windows and writes it to the sheet
                    if property_set.Name == "PSet_Revit_Type_Dimensions":
                       for property in property_set.HasProperties:
                            if property.Name == "Height":
                                height = property.NominalValue.wrappedValue
                                height = round(height,1)
                                sheet.write(row,column+2,height)
                                
                    #Sort by the name of the propertySet and gets the width of windows, calculate the area and write it to the sheet
                    if property_set.Name == "PSet_Revit_Type_Dimensions":
                        for property in property_set.HasProperties:
                            if property.Name == "Width":
                                width = property.NominalValue.wrappedValue
                                width = round(height,1)
                                sheet.write(row,column+3,width)
                                area = width*height
                                arealist.append(area)
                                sheet.write(row,column+4,area)
                                row+=1
                                
                    #Sort by the name of the propertySet and gets the level of windows and writes it to the sheet          
                    if property_set.Name == "PSet_Revit_Constraints":
                        for property in property_set.HasProperties:
                            if property.Name == "Level":
                                window_level = property.NominalValue.wrappedValue
                                sheet.write(row,column,window_level)
                         
                        #makes a list that is getting used later on that stores the orientation of the windows.
                        for product in productList:
                            if product in productList:
                                orientationlist = []
                                i=1
                                
                                #loops for every window and get its placement in vector form. It is then transposed.
                                for entity in model.by_type(product):
                                    m = local_placement(entity.ObjectPlacement)
                                    entity_matrix = np.matrix([
                                        [m[0][0], m[1][0], m[2][0], 0],
                                        [m[0][1], m[1][1], m[2][1], 0],
                                        [m[0][2], m[1][2], m[2][2], 0],
                                        [m[0][3], m[1][3], m[2][3], 1]])
                                    entity_matrix.transpose()
                                    
                                    #if-else statement is used to correspond the vectors into orientations and write them to the sheet
                                    if m[0][0] > 0:
                                        Orientation = "North"
                                        orientationlist.append("North")
                                    elif m[0][0] < 0:
                                        Orientation = "South"
                                        orientationlist.append("South")
                                    elif m[1][0] > 0:
                                        Orientation = "West"
                                        orientationlist.append("West")
                                    else:
                                        Orientation = "East"
                                        orientationlist.append("East")
                                    sheet.write(i,column+1,str(Orientation)) 
                                    i+=1
                                    
        #we take the length of the area list and loops it for all of the orientations. The solar radiation factor is then multiplied on.
        for i in range(len(arealist)):
            if orientationlist[i] == "North":
                solarirradiance = round(arealist[i]*175,0)
            elif orientationlist[i] == "South":
                solarirradiance = round(arealist[i]*790,0)
            else:
                solarirradiance = round(arealist[i]*710,0)
            sheet.write(i+1,column+5,float(solarirradiance)) 
          
makeASheet('IfcWindow')

# it is important to close the workbook afterwards
workbook.close()

#When this print message gets shown to the console an excel file has succesfully been written.
print("Excel sheet succesfully created! Check your output folder")
